### ISSUE TITLE
Eager loading extension doesn't allow change context dynamically

### ISSUE THREAD

**Issue by CvekCoding** [api-platform/core#2157]
In my graphql-based application I would like to change normalization context dynamically while `EagerLoadingExtension` builds query. E.g. i want to have different `Groups` for different queries (I can distinguish between them by request names).

As I see, you deliberately skip the place where I could connect my `ContextBuilder` and change the context (I already have `Groups` and `Attributes` in my entity):
https://github.com/api-platform/core/blob/797cea682956871c0a848275bbaeabf1339ad615/src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php#L105-L108

 Why is this done? Or did I choose the wrong way to solve my problem?

Thanks.

**Comment by Siregacvek** [comment]
@soyuka 
> normalization group per request

what do you mean? I don't see [here](https://api-platform.com/docs/core/serialization) this option. I see the following ways:
- set context in `ApiResourse()`. It's a global setting and doesn't fit my needs.
- set context per operation (query, delete...). Global again.
- change context in `ContextBuilder`. Looks good for me, but `createFromRequest` method never invoke for GraphQl.
- register my own normalizer, where I can set my own context. It invokes in GraphQl application, but only after `EagerLoading` extension - when sql query already buit and data fetched.

As far as I see, `eagerLoadingExtension` leverages this method to find the proper `Groups`:
https://github.com/api-platform/core/blob/797cea682956871c0a848275bbaeabf1339ad615/src/Metadata/Property/Factory/SerializerPropertyMetadataFactory.php#L142-L164

May be this will be good to be able to decorate/override this method to pass the groups dynamically?

**Comment by soyuka** [comment]
> set context per operation (query, delete...). Global again.

Okay my bad I confused per "operation" vs per "request".

Indeed EagerLoading is using groups to build the query. Isn't there a way we could fix the EagerLoadingExtension to use dynamic groups?

**Comment by Siregacvek** [comment]
@soyuka, I thought about this regarding GraphQl.

I think that because of the most important GraphQl-specific feature - user can select fields he wants to fetch, I believe there is a sense to consider all the fields specified in graphQl query as a dynamic group - to pull them eagerly. After all, the query itself describes the required fields, in my mind this is equivalent to the fact that the user has created a serializer group and wants to pull them out as quickly as possible (eager). Why not make all the necessary joins on them automatically and pull them all out with one query even without of requiring developer to build his own context?

What do you think?

**Comment by soyuka** [comment]
Yes indeed, I've built the EagerLoadingExtension with that in mind, it's why I based the query upon serialization groups. They indeed describe what we need.

I'm really not familiar with GraphQL but if possible it would definitely be a good feature. We must be careful that GraphQl queries don't interfere too much with groups though. 

If you're up for it, we could add such a feature to the core, with a possibility to disable it and still use groups and graphql independently.

@teohhanhui @alanpoulain @dunglas opinion on this?

**Comment by dunglas** [comment]
The problem is: eager loading is not always the fastest solution. Often, it is faster to do several queries than a single complex one. It’s why I disabled this behavior with GraphQL. We may change this default behavior, but we need to do it very carefully (i.e allowing to disabling eager loading easily). Would you want to work on this?

**Comment by Siregacvek** [comment]
I implemented this behavior (eager loading of all the fields from query and ignoring Groups for now) and dramatically reduced amount of DB queries in my application.

But then I faced the other problem here:
https://github.com/api-platform/core/blob/797cea682956871c0a848275bbaeabf1339ad615/src/GraphQl/Serializer/ItemNormalizer.php#L48-L54

The problem is in `$data[self::ITEM_KEY] = serialize($object)`. Due to the fact that eager-loaded objects are much bigger, serialization (and then - deserialization in Webonyx) becomes very expensive. As far as I understand here we don't need the real full object - empty object of the proper type with the real id will be enough. So instead of serializing the real object (which can include relations) here I instantiate empty object, copy into it `id` from real object, and serialize it.

In my application this works fine and time of query execution was reduced in about 5-20 times (depending on query complexity and amount of entities in collections).

**Comment by dunglas** [comment]
Would you mind opening a PR?

**Comment by Siregacvek** [comment]
@dunglas , sorry for late answer. Yes, of course I'll do this. I'll try to do this on weekend.

**Comment by soyuka** [comment]
> sorry for late answer

No rush, it's open source :). Take all the time you need, we look forward to review this improvement!
